service/history/workflow/retry_test.go

```diff
@@ -153,241 +152,192 @@ func Test_IsRetryable(t *testing.T) {
-			serverFailure,
-			ai.RetryNonRetryableErrorTypes,
+func Test_nextBackoffInterval(t *testing.T) {
+	now, _ := time.Parse(time.RFC3339, "2018-04-13T16:08:08+00:00")
+	expirationIn := func(t time.Duration) *timestamppb.Timestamp { return timestamppb.New(now.Add(t)) }
+	initInterval := func(t time.Duration) *durationpb.Duration { return durationpb.New(t) }
+	maxInterval := initInterval
+
+	t.Run("first attempt should use initial backoff", func(t *testing.T) {
+		currentAttempt := int32(1)
+		initialDelay := 5 * time.Second
+		interval, retryState := nextBackoffInterval(
+			doNotCare(now),
+			currentAttempt,
+			5,
+			initInterval(initialDelay),
+			doNotCare(maxInterval(10*time.Second)),
+			doNotCare(expirationIn(30*time.Second)),
+			doNotCare[float64](2),
 		)
-		a.Equal(time.Millisecond*4, interval)
-		a.Equal(enumspb.RETRY_STATE_IN_PROGRESS, retryState)
+		assert.Equal(t, initialDelay, interval)
+		assert.Equal(t, enumspb.RETRY_STATE_IN_PROGRESS, retryState)
 	})
 
-	t.Run("test non-retryable error", func(t *testing.T) {
-		ai.Attempt = 4
-		serverFailure = failure.NewServerFailure("some non-retryable server failure", true)
-		interval, retryState := getBackoffInterval(
-			now,
-			ai.Attempt,
-			ai.RetryMaximumAttempts,
-			ai.RetryInitialInterval,
-			ai.RetryMaximumInterval,
-			ai.RetryExpirationTime,
-			ai.RetryBackoffCoefficient,
-			serverFailure,
-			ai.RetryNonRetryableErrorTypes,
+	t.Run("negative or 0 attempt should be treated as first", func(t *testing.T) {
+		currentAttempt := int32(-1)
+		maxAttempts := int32(5)
+		initialDelay := 5 * time.Second
+		interval, retryState := nextBackoffInterval(
+			doNotCare(now),
+			currentAttempt,
+			maxAttempts,
+			doNotCare(initInterval(initialDelay)),
+			doNotCare(maxInterval(10*time.Second)),
+			doNotCare(expirationIn(30*time.Second)),
+			doNotCare[float64](2),
 		)
-		a.Equal(backoff.NoBackoff, interval)
-		a.Equal(enumspb.RETRY_STATE_NON_RETRYABLE_FAILURE, retryState)
-
-		serverFailure = failure.NewServerFailure("good-reason", false)
-
-		interval, retryState = getBackoffInterval(
-			now,
-			ai.Attempt,
-			ai.RetryMaximumAttempts,
-			ai.RetryInitialInterval,
-			ai.RetryMaximumInterval,
-			ai.RetryExpirationTime,
-			ai.RetryBackoffCoefficient,
-			serverFailure,
-			ai.RetryNonRetryableErrorTypes,
+		assert.Equal(t, initialDelay, interval)
+		assert.Equal(t, enumspb.RETRY_STATE_IN_PROGRESS, retryState)
+	})
+
+	t.Run("n-th retry should be initial * backoff^(n - 1)", func(t *testing.T) {
+		initialDelay := 2 * time.Second
+		attempt := int32(4)
+		maxAttempts := int32(5)
+		interval, retryState := nextBackoffInterval(
+			doNotCare(now),
+			attempt,
+			maxAttempts,
+			initInterval(initialDelay),
+			doNotCare(maxInterval(200*time.Second)),
+			doNotCare(expirationIn(600*time.Second)),
+			3,
 		)
-		a.Equal(time.Millisecond*8, interval)
-		a.Equal(enumspb.RETRY_STATE_IN_PROGRESS, retryState)
+		assert.Equal(t, initialDelay*pow(3, int32(attempt)-1), interval)
```

Comment:
Usually I'd want the expected value to be a literal because nothing is clearer than that, but given this is a mathematical expression this is fine.